database/database.go

```diff
@@ -276,11 +284,11 @@ func (r databaseRepository) SearchBlobMetadata(ctx context.Context, tags map[str
 	return results, nil, nil
 }
 
-func (r databaseRepository) GetPageOfExpiredStagedBlobMetadata(ctx context.Context, olderThan time.Time) ([]core.BlobKey, error) {
+func (r databaseRepository) GetPageOfExpiredBlobMetadata(ctx context.Context, olderThan time.Time) ([]core.BlobKey, error) {
 	rows, err := r.db.
 		Model(blobMetadata{}).
 		Select(`subject, id`).
-		Where(`staged = ? AND created_at < ?`, true, olderThan.UnixMilli()).
+		Where(`(staged = ? AND created_at < ?) OR (expires_at < ?)`, true, olderThan.UnixMilli(), olderThan.UnixMilli()).
```

Comment:
Some databases don't optimize some queries with ORs very well, and fall back to table scans. Have you checked the postgres query plan for this? Another approach is to do a `UNION ALL` with two queries.

Reply:
I had not checked, but I've had a look: 

```sql
EXPLAIN ANALYZE SELECT subject, id FROM "blob_metadata" WHERE (staged = true AND created_at < 1637748781707) OR expires_at < 1637748781707 ORDER BY created_at ASC LIMIT 200;
```
```
QUERY PLAN                                                                                                           
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Limit  (cost=4.29..4.29 rows=1 width=42) (actual time=0.047..0.048 rows=0 loops=1)                                   
  ->  Sort  (cost=4.29..4.29 rows=1 width=42) (actual time=0.047..0.047 rows=0 loops=1)                              
        Sort Key: created_at                                                                                         
        Sort Method: quicksort  Memory: 25kB                                                                         
        ->  Seq Scan on blob_metadata  (cost=0.00..4.28 rows=1 width=42) (actual time=0.028..0.028 rows=0 loops=1)   
              Filter: ((staged AND (created_at < '1637748781707'::bigint)) OR (expires_at < '1637748781707'::bigint))
              Rows Removed by Filter: 152                                                                            
Planning Time: 0.075 ms                                                                                              
Execution Time: 0.060 ms                                                                                             
(9 rows)
```

```sql
EXPLAIN ANALYZE SELECT subject, id FROM "blob_metadata" WHERE staged = true AND created_at < 1637750129345 UNION ALL SELECT subject, id FROM "blob_metadata" WHERE expires_at < 1637750129345
```
```
QUERY PLAN                                                                                                                  
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Append  (cost=0.00..7.83 rows=2 width=34) (actual time=0.030..0.030 rows=0 loops=1)                                         
  ->  Seq Scan on blob_metadata  (cost=0.00..3.90 rows=1 width=34) (actual time=0.019..0.019 rows=0 loops=1)                
        Filter: (staged AND (created_at < '1637750129345'::bigint))                                                         
        Rows Removed by Filter: 152                                                                                         
  ->  Seq Scan on blob_metadata blob_metadata_1  (cost=0.00..3.90 rows=1 width=34) (actual time=0.011..0.011 rows=0 loops=1)
        Filter: (expires_at < '1637750129345'::bigint)                                                                      
        Rows Removed by Filter: 152                                                                                         
Planning Time: 0.075 ms                                                                                                     
Execution Time: 0.040 ms                                                                                                    
(9 rows)
```
It looks like the UNION query is indeed more efficient. I'm not sure it'll matter in any appreciable way - both queries are pretty light as far as I can tell - but I'm happy to go with the UNION ALL approach.

Reply:
The remaining problem is that you don't have any kind of index to make the query for expired records efficient. I have ~ 10 million rows in the `blob_metadata` table.

This query uses a filtered index:
```
EXPLAIN ANALYZE 
SELECT subject, Id 
FROM blob_metadata
WHERE staged = true and created_at < 1637786288230

QUERY PLAN
Index Scan using staged on blob_metadata  (cost=0.12..8.14 rows=1 width=35) (actual time=0.017..0.017 rows=0 loops=1)
  Index Cond: (created_at < '1637786288230'::bigint)
Planning Time: 0.144 ms
Execution Time: 0.040 ms
```

This one does not:
```
EXPLAIN ANALYZE
SELECT subject, Id
FROM blob_metadata
WHERE expires_at < 1637786288230

Gather  (cost=1000.00..194822.96 rows=1 width=35) (actual time=580.996..584.048 rows=0 loops=1)
  Workers Planned: 2
  Workers Launched: 2
  ->  Parallel Seq Scan on blob_metadata  (cost=0.00..193822.86 rows=1 width=35) (actual time=542.972..542.973 rows=0 loops=3)
        Filter: (expires_at < '1637786288230'::bigint)
        Rows Removed by Filter: 3489451
Planning Time: 0.098 ms
JIT:
  Functions: 12
  Options: Inlining false, Optimization false, Expressions true, Deforming true
  Timing: Generation 2.476 ms, Inlining 0.000 ms, Optimization 2.073 ms, Emission 21.643 ms, Total 26.191 ms
Execution Time: 585.222 ms
```
But I can fix that with this index:
```
CREATE INDEX expires on blob_metadata 
(
    expires_at
)
WHERE expires_at IS NOT NULL
```

Now the same query is handled much more efficiently:
```
EXPLAIN ANALYZE 
SELECT subject, Id 
FROM blob_metadata
WHERE staged = true and created_at < 1637786288230

QUERY PLAN
Index Scan using expires on blob_metadata  (cost=0.12..8.14 rows=1 width=35) (actual time=0.007..0.028 rows=0 loops=1)
  Index Cond: (expires_at < '1637786288230'::bigint)
Planning Time: 0.125 ms
Execution Time: 0.126 ms
```

Now we can compare the two original queries (note that you don't need an ORDER BY).
```
EXPLAIN ANALYZE 
SELECT subject, Id 
FROM blob_metadata
WHERE staged = true AND created_at < 1637786288230 OR expires_at < 1637786288230

QUERY PLAN
Bitmap Heap Scan on blob_metadata  (cost=8.27..12.28 rows=1 width=35) (actual time=0.019..0.021 rows=0 loops=1)
  Recheck Cond: (((created_at < '1637786288230'::bigint) AND staged) OR (expires_at < '1637786288230'::bigint))
  ->  BitmapOr  (cost=8.27..8.27 rows=1 width=0) (actual time=0.016..0.017 rows=0 loops=1)
        ->  Bitmap Index Scan on staged  (cost=0.00..4.13 rows=1 width=0) (actual time=0.014..0.014 rows=0 loops=1)
              Index Cond: (created_at < '1637786288230'::bigint)
        ->  Bitmap Index Scan on expires  (cost=0.00..4.13 rows=1 width=0) (actual time=0.001..0.001 rows=0 loops=1)
              Index Cond: (expires_at < '1637786288230'::bigint)
Planning Time: 0.300 ms
Execution Time: 0.124 ms
```
And
```
EXPLAIN ANALYZE 
SELECT subject, Id 
FROM blob_metadata
WHERE staged = true AND created_at < 1637786288230
UNION ALL 
SELECT subject, Id 
FROM blob_metadata
WHERE expires_at < 1637786288230

QUERY PLAN
Append  (cost=0.12..16.31 rows=2 width=35) (actual time=0.689..0.690 rows=0 loops=1)
  ->  Index Scan using staged on blob_metadata  (cost=0.12..8.14 rows=1 width=35) (actual time=0.676..0.677 rows=0 loops=1)
        Index Cond: (created_at < '1637786288230'::bigint)
  ->  Index Scan using expires on blob_metadata blob_metadata_1  (cost=0.12..8.14 rows=1 width=35) (actual time=0.009..0.009 rows=0 loops=1)
        Index Cond: (expires_at < '1637786288230'::bigint)
Planning Time: 0.551 ms
Execution Time: 1.016 ms
```
Looks like the OR is more efficient in the end.